File: scrappers/biotech.py
import re
import scrapy
from scrapy.http import Request
from scrapy.crawler import CrawlerProcess
from .utils.connection import make_mongo_conn
from .utils.parser import text_to_number
from .dto.product import Product
from .utils.pricecalculator import PriceCalculator

class Biotech(scrapy.Spider):
    name = 'biotech'

    def start_requests(self):
        self.connection = make_mongo_conn()
        self.categories = {}
        self.base_url = 'https://biotechchile.cl'
        self.calculator = PriceCalculator(self.connection)

        yield Request(self.base_url, callback=self._fetch_categories)

    def _fetch_categories(self, response):
        links = [
            f'{self.base_url}{link}' for link in response.css('.container-fluid.fondo_dos.header-desktop .menu-item-wrap a::attr(href)').getall()
        ]

        for link in links:
            self._add_category(link)

            for i in range(1, 5):
                new_url = f'{link}/page/{i}'
                yield Request(new_url, callback=self._parse_list)

    def _add_category(self, url):
        tokens = url.split('/')[-1].split('-')
        pk = tokens.pop()
        category_name = '-'.join(tokens)
        self.categories[pk] = category_name

    def _get_category_from_url(self, url):
        result = re.search(r'category=(?P<category_id>\d+)', url)
        category_name = self.categories[result.group('category_id')]
        return category_name

    def _parse_list(self, response):
        links = response.css('.card.oe_product_cart a::attr(href)').getall()
        new_links = [f'https://biotechchile.com{url}' for url in links if url != '#']

        for link in new_links:
            yield Request(link, callback=self._parse_detail)

    def _parse_detail(self, response):
        category = self._get_category_from_url(response.url)
        title = response.css('#product_details > h1::text').get()
        price = response.css('.oe_price > .oe_currency_value::text').get()
        image = response.css('.carousel.slide img::attr(src)').get()
        stock = response.css('.availability_messages.o_not_editable > div').get()

        output = Product()
        output.title = title
        output.price = text_to_number(price)
        output.revenue_price = self.calculator.calculate_price_with_revenue(output.price)
        output.refer_url = response.url
        output.image = image
        output.brand = ''
        output.description = ''
        output.sku = ''
        output.platform_source = 'biotech'
        output.stock = text_to_number(stock)
        output.add_category(category)

        self.connection.products.insert_one(output.to_serializable())


process = CrawlerProcess(settings={})
process.crawl(Biotech)
process.start()

File: scrappers/biotech.js
const pino = require('pino');
const makeMongoConn = require('./utils/mongoConn');
const makeNewBrowser = require('./utils/makeBrowser');
const parseNumber = require('./utils/numberParser');
const container = require('../src/util/container');

const logger = pino();
const pricingService = container.get('pricingService');

const main = async () => {
  const browser = await makeNewBrowser();
  const baseURL = 'https://biotechchile.cl';

  logger.info('going to ... %s', baseURL);

  const categoryLinks = await fetchCategories(browser, baseURL);
  console.log('category links', categoryLinks);

  do {
    const link = categoryLinks.shift();
    const url = `${baseURL}${link}`;
    logger.info('url created -> %s', url);
    const { totalPages, category } = await fetchCatAndQuantity(browser, url);

    for (let i = 0; i <= totalPages; i++) {
      const pageNumber = i + 1;
      const plp = `${url}/page/${pageNumber}`;
      const pdpItems = await getProductsLink(browser, plp);

      while (pdpItems.length > 0) {
        const pdpSection = pdpItems.shift();
        const pdpUrl = `${baseURL}${pdpSection}`;
        const results = await getInfo(browser, pdpUrl, category);

        if (!results.title) {
          continue;
        }

        await insertDB(results);
      }
    }
  } while (categoryLinks.length > 0);

  browser.close();
};

const getProductsLink = async (browser, url) => {
  const page = await browser.newPage();
  await page.goto(url);

  const productsLink = await page.evaluate(() => {
    return Array.from(document.querySelectorAll('.d-block.h-100'))
      .map(href => href.getAttribute('href'));
  });

  await page.close();
  return productsLink;
};

const fetchCatAndQuantity = async (browser, url) => {
  const page = await browser.newPage();
  await page.goto(url);

  const totalPagesAndCategory = await page.evaluate(() => {
    const lastPage = Array.from(
      document.querySelectorAll('.products_pager.form-inline.flex-md-nowrap .page-item'))
      .slice(1,5)
      .map(content => content.textContent.trim());

    const category = window.location.href.split('/').slice(-1)[0].split('-')[0];

    return {
      totalPages: Number.parseInt(lastPage.slice(-1)[0]),
      category
    };
  });

  await page.close();
  return totalPagesAndCategory;
};

const fetchCategories = async (browser, url) => {
  const page = await browser.newPage();
  await page.goto(url);

  const results = await page.evaluate(() => {
    return Array.from(
        document.querySelectorAll('.container-fluid.fondo_dos.header-desktop .menu-item-wrap a')
      ).map(href => href.getAttribute('href'));
  });

  await page.close();
  return results;
};

const getInfo = async (browser, url, category) => {
  const page = await browser.newPage();
  page.on('console', (message) => {
    if (message.text().startsWith('https')) {
      logger.error('error on page %s', message.text());
    }
  });

  logger.info('inspecting %s', url);

  await page.goto(url);
  await page.waitForTimeout(3000);

  const data = await page.evaluate((category, url) => {
    try {
      const title = document.querySelector('#product_details > h1').textContent;
      const price = document.querySelector('.oe_price > .oe_currency_value').textContent;
      const image = document.querySelector('.carousel.slide img').getAttribute('src');
      const stock = document.querySelector('.availability_messages.o_not_editable > div').textContent.trim();
      const date = new Date().toISOString();

      return {
        title,
        stock,
        category,
        image,
        sku: '',
        brand: '',
        platformSource: 'biotech',
        description: '',
        createdAt: date,
        referUrl: url,
        internetPrice: price,
      };
    } catch (err) {
      console.error(url);
      return {};
    }
  }, category, page.url());

  data.price = parseNumber(data.internetPrice);
  data.stock = parseNumber(data.stock);
  data.revenuePrice = pricingService.calculatePriceWithRevenue(data.internetPrice);

  await page.close();
  return data;
};

const insertDB = async (entity) => {
  logger.info('inserting entity %j', entity);
  const connection = await makeMongoConn();
  const db = connection.db('dentshop');
  const collection = db.collection('products');

  await collection.insertOne(entity);
  await connection.close();

  return Promise.resolve();
};

main();
